FEATURE: Support TYPO3 CMS 7 to 8

DanielSiepmann opened the pull request:
In order to make a first release and further development easier,
we support both TYPO3 CMS versions with one code base.

We therefore adjust tests and implementations.

DanielSiepmann committed:
TASK: Cleanup differences

We have some small differences between develop and support/76 branch.
This differences are just CGL and should not exist.

DanielSiepmann committed:
TASK: Migrate existing 7.6 features

We had some features in 7.6 support which we didn't merge up yet.
Mostly very small bug fixes or more helpful logging and processing of
elasticsearch options.

But also adding images of content elements while indexing pages.

DanielSiepmann committed:
TASK: Do not import unnecessary namespace

As imported class is on same level as we are, we do not need to import
the class.

DanielSiepmann committed:
TASK: Support tests for TYPO3 CMS 7.6

Check which version to run and switch TYPO3 bootstrapping.
Also allow tests to run with TYPO3 CMS 7.6 again.

DanielSiepmann committed:
TASK: Provide compatible TypoScriptService for both TYPO3 versions

DanielSiepmann committed:
TASK: Provide already used logger

As we use the logger, we should inject it.

DanielSiepmann committed:
TASK: Fix call to non existing method

DanielSiepmann committed:
TASK: Make extension compatible with CMS 7.6

DanielSiepmann committed:
TASK: Use imported interface

DanielSiepmann committed:
Merge remote-tracking branch 'origin/develop' into feature/support-typo3-7-to-8

Conflicts:
    Tests/Functional/Hooks/DataHandler/ProcessesAllowedTablesTest.php
    Tests/Unit/Domain/Index/TcaIndexer/TcaTableServiceTest.php

DanielSiepmann committed:
TASK: Keep old logic to make changes smaller

DanielSiepmann committed:
TASK: Remove unused comment

DanielSiepmann committed:
TASK: Fix merge conflict issue

DanielSiepmann committed:
TASK: Make tests compatible with TYPO3 CMS 7.6

As there is no doctrine, we receive values as string, also for uid.

DanielSiepmann committed:
TASK: Add necessary phpdoc for extbase

As extbase does not reflect PHP source but phpdoc, we have to define the
type to use for mapping as phpdoc.

justusmoroni commented on Classes/Domain/Index/TcaIndexer/PagesIndexer.php:
```diff
@@ -60,28 +67,73 @@ protected function prepareRecord(array &$record)
+
+    protected function getContentElementImages(int $uidOfContentElement) : array
+    {
+        return $this->fetchSysFileReferenceUids($uidOfContentElement, 'tt_content', 'image');
```
`image` is not the only field where FSC (Fluid Styled Content) can save them, right?

I think `TextMedia` uses the field `assets`.

DanielSiepmann replied:
I'll keep it as it s and will change it once we need to.
Otherwise this will never end and we will fix existing issues all the time.

justusmoroni commented on Classes/Domain/Index/TcaIndexer/TcaTableService76.php:
```diff
@@ -0,0 +1,379 @@
+            $this->tca['ctrl']['cruser_id'],
+            $this->tca['ctrl']['fe_cruser_id'],
+            $this->tca['ctrl']['fe_crgroup_id'],
+            $this->tca['ctrl']['languageField'],
```
Wasn't this removed for the 8.7 part? Shouldn't this be removed here too?

https://github.com/Codappix/search_core/pull/132/files

justusmoroni replied:
Maybe you could move that part to an trait, so that it is not defined in 2 places?

You could do that for other parts as well.

justusmoroni replied:
Why did you not extend the normal TcaTableService and only change the stuff that was needed?

DanielSiepmann replied:
I want to get rid of the old 76 implementations as soon as possible. I'll not add further code that will produce another cleanup.

Instead I've added a test for language handling and removed the line.

justusmoroni commented on Classes/Domain/Index/TcaIndexer/TcaTableService76.php:
```diff
@@ -0,0 +1,379 @@
+
+    public function getTableClause() : string
+    {
+        if ($this->tableName === 'pages') {
```
Why is this part different from the 8.7 part?

https://github.com/Codappix/search_core/blob/feature/support-typo3-7-to-8/Classes/Domain/Index/TcaIndexer/TcaTableService.php#L118

DanielSiepmann replied:
Because in 8 we have doctrine and use `getJoins` to fetch the join information

justusmoroni commented on Tests/Unit/Domain/Index/TcaIndexer/TcaTableServiceTest.php:
```diff
@@ -82,23 +98,27 @@ public function doUsePlainQueryIfNoAdditionalWhereClauseIsDefined()
-        );
+        $this->subject->getRecord(10);
+
+        // $whereClause = $this->subject->getWhereClause();
```
Remove.

justusmoroni replied:
Or keep - you decide. But we usually don't keep these commented code lines.

DanielSiepmann replied:
I've marked tests as incomplete and reverted to comment code out.

justusmoroni commented on Tests/Unit/Domain/Index/TcaIndexer/TcaTableServiceTest.php:
```diff
@@ -141,16 +161,16 @@ public function allConfiguredAndAllowedTcaColumnsAreReturnedAsFields()
         );
         $this->inject($subject, 'logger', $this->getMockedLogger());
 
-        $this->assertSame(
```
Remove or keep - you decide. But we usually don't keep these commented code lines.

DanielSiepmann replied:
I've marked tests as incomplete and reverted to comment code out.

DanielSiepmann committed:
TASK: Keep language information for all TYPO3 Versions

Keep code consistent. Fetch language field for both TYPO3 versions.
To make sure we do not mess up, add test case.

DanielSiepmann committed:
TASK: Add missing fixture

DanielSiepmann committed:
TASK: Do not comment out code for non working tests

Mark tests as non working for now.

DanielSiepmann committed:
BUGFIX: Workaround doctrine limitation

Doctrine will not implode our array, so we have to do

DanielSiepmann committed:
TASK: Remove unnecessary fixture
